ADApp/ntndArrayConverterSrc/ntndArrayConverter_pvxs.cpp

```diff
@@ -0,0 +1,447 @@
+    NDAttrSource_t sourceType = (NDAttrSource_t) attribute["sourceType"].as<int32_t>();
+    auto value = attribute["value"].as<std::string>();
+
+    NDAttribute *attr = new NDAttribute(name.c_str(), desc.c_str(), sourceType, source.c_str(), NDAttrDataType_t::NDAttrString, (void*)value.c_str());
+    dest->pAttributeList->add(attr);
+}
+
+void NTNDArrayConverter::toUndefinedAttribute (NDArray *dest, pvxs::Value attribute)
+{
+    auto name = attribute["name"].as<std::string>();
+    auto desc = attribute["descriptor"].as<std::string>();
+    auto source = attribute["source"].as<std::string>();
+    NDAttrSource_t sourceType = (NDAttrSource_t) attribute["sourceType"].as<int32_t>();
+
+    NDAttribute *attr = new NDAttribute(name.c_str(), desc.c_str(), sourceType, source.c_str(), NDAttrDataType_t::NDAttrUndefined, NULL);
+    dest->pAttributeList->add(attr);
+}
+
+void NTNDArrayConverter::toAttributes (NDArray *dest)
+{
+    auto attributes = m_value["attribute"].as<pvxs::shared_array<const pvxs::Value>>();
+    for (int i=0; i<attributes.size(); i++) {
+        pvxs::Value value = attributes[i]["value"];
+        switch (attributes[i]["value->"].type().code) {
+            // use indirection on Any container to get specified type
+            case pvxs::TypeCode::Int8:        toAttribute<int8_t>    (dest, attributes[i]); break;
+            case pvxs::TypeCode::UInt8:       toAttribute<uint8_t>   (dest, attributes[i]); break;
+            case pvxs::TypeCode::Int16:       toAttribute<int16_t>   (dest, attributes[i]); break;
+            case pvxs::TypeCode::UInt16:      toAttribute<uint16_t>  (dest, attributes[i]); break;
+            case pvxs::TypeCode::Int32:       toAttribute<int32_t>   (dest, attributes[i]); break;
+            case pvxs::TypeCode::UInt32:      toAttribute<uint32_t>  (dest, attributes[i]); break;
+            case pvxs::TypeCode::Int64:       toAttribute<int64_t>   (dest, attributes[i]); break;
+            case pvxs::TypeCode::UInt64:      toAttribute<uint64_t>  (dest, attributes[i]); break;
+            case pvxs::TypeCode::Float32:     toAttribute<float_t>   (dest, attributes[i]); break;
+            case pvxs::TypeCode::Float64:     toAttribute<double_t>  (dest, attributes[i]); break;
+            case pvxs::TypeCode::String:      toStringAttribute      (dest, attributes[i]); break;
+            case pvxs::TypeCode::Null:        toUndefinedAttribute   (dest, attributes[i]); break;
+            default: throw std::runtime_error("invalid value data type");
+        }
+    }
+}
+
+template <typename dataType>
+void NTNDArrayConverter::fromValue(NDArray *src) {
+    NDArrayInfo_t arrayInfo;
+    src->getInfo(&arrayInfo);
+
+    m_value["compressedSize"] = src->compressedSize;
+    m_value["uncompressedSize"] = arrayInfo.totalBytes;
+    std::string fieldName = m_fieldNameMap[typeid(dataType)];
+    auto arrayType = m_arrayTypeMap[typeid(dataType)];
+    const auto val = pvxs::detail::copyAs(
```

Comment:
The `pvxs::detail` namespace is not part of the public API.  Please do not use members directly.

In the `dtype==stype` case, and in the short term, I think you would do better to `pvxs::allocArray()` and then `memcpy()`.

In the longer term, I think it would be interesting to see if this copy could be avoided by building a `pvxs::shared_array` around an `NDArray` reference.